src/uu/factor/src/miller_rabin.rs

```diff
@@ -0,0 +1,88 @@
+use crate::numeric::*;
+
+// Small set of bases for the Miller-Rabin prime test, valid for all 64b integers;
+//  discovered by Jim Sinclair on 2011-04-20, see miller-rabin.appspot.com
+#[allow(clippy::unreadable_literal)]
+const BASIS: [u64; 7] = [2, 325, 9375, 28178, 450775, 9780504, 1795265022];
+
+#[derive(Eq, PartialEq)]
+pub(crate) enum Result {
+    Prime,
+    Pseudoprime,
+    Composite(u64),
+}
+
+impl Result {
+    pub(crate) fn is_prime(&self) -> bool {
+        *self == Result::Prime
+    }
+}
+
+// Deterministic Miller-Rabin primality-checking algorithm, adapted to extract
+// (some) dividers; it will fail to factor strong pseudoprimes.
+#[allow(clippy::many_single_char_names)]
+pub(crate) fn test<A: Arithmetic>(n: u64) -> Result {
+    use self::Result::*;
+
+    if n < 2 {
+        return Pseudoprime;
+    }
+    if n % 2 == 0 {
+        return if n == 2 { Prime } else { Composite(2) };
+    }
+
+    // n-1 = r 2ⁱ
+    let i = (n - 1).trailing_zeros();
+    let r = (n - 1) >> i;
+
+    for a in BASIS.iter() {
+        let a = a % n;
+        if a == 0 {
+            break;
+        }
+
+        // x = a^r mod n
+        let mut x = A::pow(a, r, n);
```

Comment:
We compute 7 times `a^r` with a fixed `r` (and different `a`s) so it may be faster to precompute an optimal exponentiation ladder rather than using the naive binary-squaring method.
(And it's worth optimising, that one call to `pow` accounts for 60% of the runtime now.)

Reply:
Did you want to work on this PR more, or merge and do further work as another PR?

Reply:
I think merging at this point makes sense.

Reply:
Also I just realised it's midnight, and I haven't eaten, so I should leave this be until at least tomorrow.

Reply:
👍